### ISSUE TITLE
Verify Telegram Auth signature

### ISSUE THREAD

**Issue by bjesuiter** [codemonument/deno_audio_logbook#22]
When receiving a telegram auth token, we have to verify it in `/auth/callback.tsx`.

## Docs 
- Telegram Auth Instructions: https://core.telegram.org/widgets/login#checking-authorization 
- Useful gist for checking the telegram signature: 
  https://gist.github.com/Pitasi/574cb19348141d7bf8de83a0555fd2dc

## Related 
- General Docs for Telegram Login Widget: https://core.telegram.org/widgets/login

**Comment by Bloodiko** [comment]
Gosh this is Hard. 
I tried to figure it out, but nobody seems to have done it with WebCrypt publicly yet. 

### Should Work

```js
    const secrets = await secretsPromise;
    const telegramToken = z.string().parse(secrets.get("TELEGRAM_TOKEN"));

    const encoder = new TextEncoder();

    // hash telegram token with sha256

    const telegramTokenHashed = await crypto.subtle.digest(
      "SHA-256",
      encoder.encode(telegramToken),
    ).then((hash) => new Uint8Array(hash));

    //debug, skip this. 
    const tokenShaHex = Array.from(new Uint8Array(signature))
      .map((b) => b.toString(16).padStart(2, "0"))
      .join("");
    console.log(tokenShaHex)

    // Telegram Token Hash is correct, confirmed with sha256sum on terminal
```
tokenShaHex is the same as `echo -n "telegramToken"|sha256sum`

### unclear Part

```js
// generate hmac from hash and authDataCheckString
    const hmac = await crypto.subtle.importKey(
      "raw",
      telegramTokenHashed,
      { name: "HMAC", hash: "SHA-256" },
      false,
      ["sign"],
    );

    const signature = await crypto.subtle.sign(
      "HMAC",
      hmac,
      encoder.encode(authDataCheckString),
    );

    const signatureHex = Array.from(new Uint8Array(signature))
      .map((b) => b.toString(16).padStart(2, "0"))
      .join("");

    if (signatureHex !== payload.data.hash) {
      console.log("signatureHex ", signatureHex);
      console.log("payload.data.hash ", payload.data.hash);
      const msg = `Problem while verifying oauth callback payload`;

      return new Response(msg, { status: 403 });
    }
```


## Parsing Usersession to match telegram data for verification. 

```js
function buildAuthString(payload: UserSession) {
  const authStringArray = [];

  authStringArray.push(`auth_date=${payload.unixAuthDate}`);
  payload.firstName && authStringArray.push(`first_name=${payload.firstName}`);
  authStringArray.push(`id=${payload.userId}`);
  payload.lastName && authStringArray.push(`last_name=${payload.lastName}`);
  payload.photoUrl && authStringArray.push(`photo_url=${payload.photoUrl}`);
  payload.username && authStringArray.push(`username=${payload.username}`);

  return authStringArray.join("\n");
}
```

Either
* Telegram uses full string even with missing parameters (like lastname, if undefined as "" ), OR 
* the HMAC verification is wonky, 

Trying to Verify the UserSession.Hash via ` echo -n "{parsedUserSessionString with \n between each line, but not at the end}"|openssl sha256 -hmac "sha256sum(TelegramToken)"` did not yield any useful results. 
Could not verify my own callback URL after succesfull login.

**Comment by bjesuiter** [comment]
The research looks good. Think we have to go over the algortihm together again to find the problem. 
I also think the problem is with encoding the Params in the thing which gets hashed. 
Probably our zod validation adds more properties into the object than should be there or otherwise messes with the hashing, so that our comparisons do not match. 

Let's see...

**Comment by bjesuiter** [comment]
![Telegram Auth Verification Algorithm excalidraw](https://user-images.githubusercontent.com/2365676/230594504-e549bc8c-f1ea-4b52-9c0d-c2e5a0700c56.svg)

**Comment by bjesuiter** [comment]
## Example code for "how to use web crypto for HMAC signing) 

[Goto TS Playground](https://www.typescriptlang.org/play?#code/DYUwLgBATiDOAOB7AdrEBlMUCWyDmEAvBAEQkDcAUAMYqyQjK0AmIUREyIA7hACogAHmACiTRKygAKAJRUaUAJ7wwiAHSwArgCMwoNdgC2SKGADSIRVMoRSUAIbcSAGhsRGLNmo8SQUkmjUMGAkMq62AN6c9oYgAFykABIAsgCCAMIuEAAW9rDZCSToiakAtABMAKwAbCQQAL7hEABm9sBoTQDaAdh4yCQAuq4yamDZjFJS2Yb21DJEAHwQEW4wCHQYWLgExAAGbraJeRBByqoJACQRpyqIEACEhMSayKzNuCDMjRBXN6r1BwgR1gJyUtw0Oj08R+1zBqghulADyeEBebw+X2cML+6i0iJAANstl28lswU0UGQoLOuMh+lgvWQ1iJSTSmSatmmsw57nEkm8fL8JESIGAwDuAHVEFBgMx7qEmnJKPURmMJlJ4PZFOL7Mx0Iz7GAKSB5oQlisyXAkKhNjh8BAANR7QEszXaxC6-V9Q3Gy4RN06vUGo0wQlEkmUNy0VCQBnekMgEWCDipKAORRqZpQRCGKRcXgAVVwYAAHKn0xqtYGvcgfTAZDJAWoZvBJtpTUttKNEJg7XgpABGaojTV6sD2UxScpYkgABlCjZZagAVohcP5QqToFaNr3to7nSzXVWPUH48aIMck3647WE0mw8SqCrRuMmbJFsso3REPpxf21mtNA93wJUVSoIA)

```ts
let responseString = "";
const encoder = new TextEncoder();

crypto.subtle.importKey(
  "raw",
  encoder.encode("secret"),
  { name: "HMAC", hash: "SHA-256" },
  false,
  ["sign"],
).then((hmac) => {
  responseString = `
    Has crypto: ${crypto !== undefined}, ${crypto}
    Has crypto.subtle: ${crypto.subtle !== undefined}, ${crypto.subtle}
    `;

  return crypto.subtle.sign(
    "HMAC",
    hmac,
    encoder.encode("Hello World!"),
  );
}).then((payloadSignature) => {
  responseString += `
        payloadSignature: ${payloadSignature}
    `;

  const signatureHex = Array.from(new Uint8Array(payloadSignature))
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");

  responseString += `
        payloadSignature as Hex: ${signatureHex}
    `;
}).then(() => {
  console.log(responseString);
});

```

**Result hash**: 6fa7b4dea28ee348df10f9bb595ad985ff150a4adfd6131cca677d9acee07dc6

## Verify with openssl

```bash 
$ echo -n "Hello World!" | openssl sha256 -hmac "secret"
SHA256(stdin)= 6fa7b4dea28ee348df10f9bb595ad985ff150a4adfd6131cca677d9acee07dc6
```

**Comment by bjesuiter** [comment]
## Extended Example Code with hashing the secret before using as an hmac key: 

[Goto TS Playground](https://www.typescriptlang.org/play?#code/DYUwLgBATiDOAOB7AdrEBlMUCWyDmEAvBAEQkDcAUAMYqyQjK0AmIUREyIA7hACogAHmACiTRKygAKAJRUaUAJ7wwiAHSwArgCMwoNc2x44YKSXQAJAIIBaAEwBWAGwkANBEYs2azxJBm0ahgwEhkZSggINTAAC0YpKUDgiwBDWBiZIgA+CABvCMjocE0oZAgg5VUNHT0QNWwAWyQoMABpEEUpAsLSKBTuN27CpPBU9NchyNzOFIaQAC5SCwBZKwBhNwgYtJjF82t7ZxIIAF8JnsiAMxTgNHOLgG0SWCNkEgBde8i5ApPwyOicWQCRiDRS1EyhBy+R6MAQdAwWFwBGIAAMhmNykoVIhFgASXIVHEQACEhGImmQrEuuBAzDOEAJRNUJwxaSxlXUWl0oHxhOxVW5tVJ5IglOptPp7iZAq5NVArJ6qPksOKpQ5OOqPLqLzwwMmS1WGy+hVB4JNkV8kh84lYZgsIGAwEQEAA6ogoMBmCTQiafpE-oD4lJ4ClFM6Usx0K8UmASiBIdChnCkKhETh8BAANRog2RUPhxCR6N62PxvkFiNRmNxmCKwrKyhDWioSC65BlmAOwQcKxQPqKNSXKCIBpSLi8ACquDAAA4+wOQ2GqyWO7WE-8LmowfAEtpExBtNFEJgM3gpABGJwyNShqNgFItKR2dwkAAMoU3PTUACtELgzFCFVChTBFT2RbNcwuQpKyLatS3XCB2W7Pl207EBu3rSJGwDG9YmDA8YUKFtYEQfRnXPUC03A-B-VOH4gA)

```ts
let responseString = "";
const encoder = new TextEncoder();

crypto.subtle.digest("SHA-256", encoder.encode("secret"))
  .then((secretHash) => {
    return crypto.subtle.importKey(
      "raw",
      secretHash,
      { name: "HMAC", hash: "SHA-256" },
      false,
      ["sign"],
    );
  })
  .then((hmac) => {
    responseString = `
    Has crypto: ${crypto !== undefined}, ${crypto}
    Has crypto.subtle: ${crypto.subtle !== undefined}, ${crypto.subtle}
    `;

    return crypto.subtle.sign(
      "HMAC",
      hmac,
      encoder.encode("Hello World!"),
    );
  }).then((payloadSignature) => {
    responseString += `
        payloadSignature: ${payloadSignature}
    `;

    const signatureHex = Array.from(new Uint8Array(payloadSignature))
      .map((b) => b.toString(16).padStart(2, "0"))
      .join("");

    responseString += `
        payloadSignature as Hex: ${signatureHex}
    `;
  }).then(() => {
    console.log(responseString);
  });

```

**Result Hash:** 69a49567d6b9d862eb301f392557ce60bedf0dff3b1589b88c3051812373af7c

## Verify with openssl

```bash 
$ echo -n "Hello World!" | openssl sha256 -mac hmac -macopt hexkey:$(echo -n "secret"|sha256sum)
SHA256(stdin)= 69a49567d6b9d862eb301f392557ce60bedf0dff3b1589b88c3051812373af7c
```

**Comment by bjesuiter** [comment]
Fixed by #31
